### DISCUSSION TITLE
[Feature]: Allow `providerImportSource` to be set in `xdm.compile` options argument via `mdx` in remix.config.js

### DISCUSSION THREAD

**Discussion by joshuaellis** [remix-run/remix#2907] [Proposals]
### What is the new or updated feature that you are suggesting?

Be able to pass `providerImportSource` in the `mdx` settings of the `remix.config.js` like so:

```js
module.exports = {
  appDirectory: 'app',
  assetsBuildDirectory: 'public/build',
  publicPath: '/build/',
  serverBuildDirectory: 'api/_build',
  ignoredRouteFiles: ['.*'],
  mdx: {
    providerImportSource: "@mdx-js/react"
  }
}
```

### Why should this feature be included?

As demonstrated in the [Advanced Configuration](https://remix.run/docs/en/v1/guides/mdx#advanced-configuration) space of the MDX sections, we can add remark & rehype plugins, which is great. However `xdm` does not supply an MDXProvider like [`@mdx-js/react`](https://mdxjs.com/docs/using-mdx/#mdx-provider?) so I cannot wrap all the docs layout in one provider to mutate the passed `components` like so:

```js
// routes/docs.tsx
const comps = {
  h1: ({ children }) => <p>{children}</p>,
}

export default function DocsLayout() {
  return (
    <>
      <Header />
      <Main>
        <MDXProvider components={comps}>
          <Outlet />
        </MDXProvider>
      </Main>
    </>
  )
}
```

You can pass components with `xdm` by [importing the MDX directly into the page](https://github.com/wooorm/xdm#mdx-content), but then I can't have a file structure like so:

```
- routes
-- index.tsx
-- docs.tsx
-- docs
--- index.mdx
```

Where I believe to do the `xdm` way I would have to set up the `components` everytime. However, setting `providerImportSource` allows us to [cross runtimes](https://github.com/wooorm/xdm#optionsproviderimportsource) and therefore enable `MDXProvider` from `@mdx-js/react`. This means I don't need to have seperate templates and content files for pages like `docs/index` or `docs/getting-started`.

I was going to create a PR straight away for this, but I thought i should float the idea first.

**Comment by axross** [comment]
@joshuaellis I'm facing the same issue. Are you still going to create a PR?

**Comment by clintonwoo** [comment]
I'm also facing the same issue. Doesn't seem like the `MDXProvider` is usable for MDX without this parameter `providerImportSource` set.

**Comment by joshuaellis** [comment]
@chaance was there a reason this feature request has been moved to a discussion? ☺️

**Comment by chaance** [comment]
@joshuaellis We are moving feature requests without current action items from issues to discussions for internal organization purposes. We are keeping our eye on both!

**Comment by ibreslauer** [comment]
It would be awesome to be able to use [MDXProvider](https://mdxjs.com/packages/react/#use) with Remix. Are there any recommended workarounds for this at the moment?

**Comment by joshuaellis** [comment]
You can use `patch-package` to specifically pass the option, but i've made a PR which would allow it to be done via the config see #3911

**Comment by AdiRishi** [comment]
Does the Remix team plan to merge PR #3911 ? Having this config option would be very helpful.

**Comment by ryanflorence** [comment]
With the forthcoming vite compiler, apps have complete control over how they want to handle non-javascript imports like MDX.

https://remix.run/docs/en/main/future/vite#add-mdx-plugin
